service/share/local_availability.go

```diff
@@ -0,0 +1,55 @@
+package share
+
+import (
+	"context"
+
+	"github.com/ipfs/go-datastore"
+	"github.com/ipfs/go-datastore/namespace"
+)
+
+var localAvailabilityPrefix = datastore.NewKey("local_availability")
+
+func rootKey(root *Root) datastore.Key {
+	return datastore.NewKey(root.String())
+}
+
+// localAvailability wraps a given Availability (whether it's light or full)
+// and stores the results of a successful sampling routine over a given Root's hash
+// to disk.
+type localAvailability struct {
+	avail Availability
+
+	ds datastore.Datastore
+}
+
+// NewLocalAvailability wraps the given Availability with an additional datastore
+// for sampling result caching.
+func NewLocalAvailability(avail Availability, ds datastore.Datastore) (Availability, error) {
+	ds = namespace.Wrap(ds, localAvailabilityPrefix)
+	return &localAvailability{
+		avail: avail,
+		ds:    ds,
```

Comment:
One IO optimization suggestion is to use https://github.com/ipfs/go-datastore/blob/master/autobatch/autobatch.go. It collects multiple IO ops and flushes them only after a certain threshold is reached.

Reply:
It also requests datastore.Batching which we do provide in DI